ansible-ipi-install/roles/node-prep/tasks/100_power_off_cluster_servers.yml

```diff
@@ -1,12 +1,24 @@
 ---
-- name: Power OFF All Nodes
-  ipmi_power:
-    name: "{{ hostvars[item]['ipmi_address'] }}"
-    user: "{{ hostvars[item]['ipmi_user'] }}"
-    password: "{{ hostvars[item]['ipmi_password'] }}"
-    port: "{{ hostvars[item]['ipmi_port'] | default(623) }}"
-    state: off
-  with_items:
-    - "{{ groups['masters'] }}"
-    - "{{ groups['workers'] }}"
+- name: Power OFF nodes
+  block:
+    - name: Create list of hosts that are going to be powered off
+      add_host:
+        groups: poweroff_hosts
+        hostname: "{{ item }}"
+        inventory_dir: "{{ hostvars[item].inventory_dir }}"
+      when: ( hostvars[item]['poweroff'] is not defined )
+        or  ( hostvars[item]['poweroff'] == true )
```

Comment:
can you confirm if this would work?
```
( hostvars[item]['poweroff']|bool )
```
Reason I mention it is is because this expression is validating a string instead of the boolean for (true or True)

Reply:
To add, I wonder if this would take care of both expression?
```
( hostvars[item]['poweroff']|bool | default(True))
```

Reply:
```
$ cat pb.yaml 
---
- name: IPI on Baremetal Installation Playbook
  hosts: localhost
  tasks:
    - name: Create list of hosts that are going to be powered off
      add_host:
        groups: poweroff_hosts
        hostname: "{{ item }}"
        inventory_dir: "{{ hostvars[item].inventory_dir }}"
      when: ( hostvars[item]['poweroff']|bool | default(True) )
      with_items:
        - "{{ groups.masters }}"
        - "{{ groups.workers }}"
    
    - name: Show hosts
      debug:
        msg: "{{ item }}"
      with_items: "{{ groups['poweroff_hosts'] }}"
```

```
$ cat inv 
[masters]
master-0 
master-1
master-2

# Worker nodes
[workers]
worker-0 poweroff=true
worker-1 poweroff=false
```

```
$ ansible-playbook -i inv pb.yaml 

PLAY [IPI on Baremetal Installation Playbook] ************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Create list of hosts that are going to be powered off] *********************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "The conditional check '( hostvars[item]['poweroff']|bool | default(True) )' failed. The error was: error while evaluating conditional (( hostvars[item]['poweroff']|bool | default(True) )): 'ansible.vars.hostvars.HostVarsVars object' has no attribute 'poweroff'\n\nThe error appears to be in '/home/edu/tmp/pb.yaml': line 5, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n    - name: Create list of hosts that are going to be powered off\n      ^ here\n"}

NO MORE HOSTS LEFT ***************************************************************************************************************************************************************************************************************************

PLAY RECAP ********************************************************************************************************
```

Reply:
```
---
- name: IPI on Baremetal Installation Playbook
  hosts: localhost
  tasks:
    - name: Create list of hosts that are going to be powered off
      add_host:
        groups: poweroff_hosts
        hostname: "{{ item }}"
        inventory_dir: "{{ hostvars[item].inventory_dir }}"
      when: ( hostvars[item]['poweroff'] is not defined )
        or  ( hostvars[item]['poweroff']|bool | default(True) )
      with_items:
        - "{{ groups.masters }}"
        - "{{ groups.workers }}"
    
    - name: Show hosts
      debug:
        msg: "{{ item }}"
      with_items: "{{ groups['poweroff_hosts'] }}"
```

This works.

Reply:
I noticed in your inventory you gave poweroff a value for both options, if you leave it out entirely does it still give the expected behavior of setting poweroff to true with the new when conditional?

Reply:
See masters, they don't have the parameter and they are added to the group.